### PR TITLE
Pseudo refactoring

### LINKED ISSUE #1305: MIPS pseudo is broken
Pseudo on mips binaries is broken.

```
│ ╎╎╎╎╎╎╎   0x08002bc8      2b104402       v0 = , unsigned  , s2 < a0
│ ────────< 0x08002bcc      03004010       if
│ ╎╎╎╎╎╎╎   0x08002bd0      00000000       
│ ╎╎╎╎╎╎╎   0x08002bd4      27b00400       s6 = ~, 0 | a0
│ ╎╎╎╎╎╎╎   0x08002bd8      ffff1224       s2 = 0 - 1
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ 0x8002bcc
│ ────────> 0x08002bdc      00000282       v0 = byte [s0 + 0]
│ ╎╎╎╎╎╎╎   0x08002be0      00001e3c       fp = 0 << 16
│ ╎╎╎╎╎╎╎   0x08002be4      0000143c       s4 = 0 << 16
│ ╎╎╎╎╎╎╎   0x08002be8      2118a002       v1 = s5
│ ╎╎╎╎╎╎╎   0x08002bec      bc02de27       fp += 0x2bc
│ ╎╎╎╎╎╎╎   0x08002bf0      e8029426       s4 += 0x2e8
│ ╎╎╎╎╎╎╎   0x08002bf4      0000173c       s7 = 0 << 16
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ +0x1bc
│ ────────> 0x08002bf8      27004010       if
│ ╎╎╎╎╎╎╎   0x08002bfc      00000000       
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ +0x114
│ ────────> 0x08002c00      21200002       a0 = s0
│ ╎╎╎╎╎╎╎   0x08002c04      2000a527       a1 = sp + 0x20              ; size
│ ╎╎╎╎╎╎╎   0x08002c08      0a01000c       call 0x428
│ ╎╎╎╎╎╎╎   0x08002c0c      2c00a3af       [sp + 0x2c] = v1
│ ╎╎╎╎╎╎╎   0x08002c10      21384000       a3 = v0
│ ╎╎╎╎╎╎╎   0x08002c14      2000a293       v0 = , unsigned  byte [sp + 0x20]
│ ╎╎╎╎╎╎╎   0x08002c18      21880702       s1 = s0 + a3
│ ╎╎╎╎╎╎╎   0x08002c1c      0800442c       a0 = , unsigned  , v0 < 8
│ ────────< 0x08002c20      10008014       if
│ ╎╎╎╎╎╎╎   0x08002c24      2c00a38f       v1 = [sp + 0x2c]
│ ╎╎╎╎╎╎╎   0x08002c28      f8ff4224       v0 -= 8
│ ╎╎╎╎╎╎╎   0x08002c2c      ff004230       v0 &= 0xff
│ ╎╎╎╎╎╎╎   0x08002c30      0b00442c       a0 = , unsigned  , v0 < 0xb
│ ────────< 0x08002c34      06008014       if
│ ╎╎╎╎╎╎╎   0x08002c38      00000000       
│ ╎╎╎╎╎╎╎   0x08002c3c      2100a293       v0 = , unsigned  byte [sp + 0x21]
│ ╎╎╎╎╎╎╎   0x08002c40      0000668e       a2 = [s3 + 0]
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ +0x3f4
│ ────────> 0x08002c44      21380000       a3 = 0
│ ────────< 0x08002c48      ad0b0008       goto 0x2eb4
│ ╎╎╎╎╎╎╎   0x08002c4c      04007326       s3 += 4
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ 0x8002c34
│ ────────> 0x08002c50      80100200       v0 <<= 2
│ ╎╎╎╎╎╎╎   0x08002c54      2110c203       v0 = fp + v0
│ ╎╎╎╎╎╎╎   0x08002c58      0000428c       v0 = [v0 + 0]
│ ╎╎╎╎╎╎╎   0x08002c5c      08004000       goto v0
```

assembly

```
│ ╎╎╎╎╎╎╎   0x08002bc8      2b104402       sltu  v0, s2, a0
│ ────────< 0x08002bcc      03004010       beqz  v0, 0x8002bdc
│ ╎╎╎╎╎╎╎   0x08002bd0      00000000       nop
│ ╎╎╎╎╎╎╎   0x08002bd4      27b00400       nor   s6, zero, a0
│ ╎╎╎╎╎╎╎   0x08002bd8      ffff1224       addiu s2, zero, -1
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ 0x8002bcc
│ ────────> 0x08002bdc      00000282       lb    v0, (s0)
│ ╎╎╎╎╎╎╎   0x08002be0      00001e3c       lui   fp, 0
│ ╎╎╎╎╎╎╎   0x08002be4      0000143c       lui   s4, 0
│ ╎╎╎╎╎╎╎   0x08002be8      2118a002       move  v1, s5
│ ╎╎╎╎╎╎╎   0x08002bec      bc02de27       addiu fp, fp, 0x2bc
│ ╎╎╎╎╎╎╎   0x08002bf0      e8029426       addiu s4, s4, 0x2e8
│ ╎╎╎╎╎╎╎   0x08002bf4      0000173c       lui   s7, 0
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ +0x1bc
│ ────────> 0x08002bf8      27004010       beqz  v0, 0x8002c98
│ ╎╎╎╎╎╎╎   0x08002bfc      00000000       nop
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ +0x114
│ ────────> 0x08002c00      21200002       move  a0, s0
│ ╎╎╎╎╎╎╎   0x08002c04      2000a527       addiu a1, sp, 0x20          ; size
│ ╎╎╎╎╎╎╎   0x08002c08      0a01000c       jal   0x428
│ ╎╎╎╎╎╎╎   0x08002c0c      2c00a3af       sw    v1, (var_2ch)
│ ╎╎╎╎╎╎╎   0x08002c10      21384000       move  a3, v0
│ ╎╎╎╎╎╎╎   0x08002c14      2000a293       lbu   v0, (var_20h)
│ ╎╎╎╎╎╎╎   0x08002c18      21880702       addu  s1, s0, a3
│ ╎╎╎╎╎╎╎   0x08002c1c      0800442c       sltiu a0, v0, 8
│ ────────< 0x08002c20      10008014       bnez  a0, 0x8002c64
│ ╎╎╎╎╎╎╎   0x08002c24      2c00a38f       lw    v1, (var_2ch)
│ ╎╎╎╎╎╎╎   0x08002c28      f8ff4224       addiu v0, v0, -8
│ ╎╎╎╎╎╎╎   0x08002c2c      ff004230       andi  v0, v0, 0xff
│ ╎╎╎╎╎╎╎   0x08002c30      0b00442c       sltiu a0, v0, 0xb
│ ────────< 0x08002c34      06008014       bnez  a0, 0x8002c50
│ ╎╎╎╎╎╎╎   0x08002c38      00000000       nop
│ ╎╎╎╎╎╎╎   0x08002c3c      2100a293       lbu   v0, (var_21h)
│ ╎╎╎╎╎╎╎   0x08002c40      0000668e       lw    a2, (s3)
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ +0x3f4
│ ────────> 0x08002c44      21380000       move  a3, zero
│ ────────< 0x08002c48      ad0b0008       j     0x2eb4
│ ╎╎╎╎╎╎╎   0x08002c4c      04007326       addiu s3, s3, 4
│ ╎╎╎╎╎╎╎   ; CODE XREF from sym.vsnprintf @ 0x8002c34
│ ────────> 0x08002c50      80100200       sll   v0, v0, 2
│ ╎╎╎╎╎╎╎   0x08002c54      2110c203       addu  v0, fp, v0
│ ╎╎╎╎╎╎╎   0x08002c58      0000428c       lw    v0, (v0)
│ ╎╎╎╎╎╎╎   0x08002c5c      08004000       jr    v0
```

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Rewritten
- MIPS
- ARM
- AVR
- 6502
- Dalvik
- z80
- tms320
- v850
- SuperH
- M68K

### CODE REVIEW

**Comment by XVilka** [comment]
Are there more representative tests for these fixes?

File: librz/core/cfile.c [hunk 1]
```diff
@@ -43,7 +43,7 @@ static void loadGP(RzCore *core) {
 		ut64 gp = rz_num_math(core->num, "loc._gp");
 		if (!gp || gp == UT64_MAX) {
 			rz_config_set(core->config, "analysis.roregs", "zero");
-			rz_core_cmd0(core, "10aes@entry0");
+			rz_core_cmd0(core, "s entry ; 10 aes");
```
Comment (XVilka) [inline librz/core/cfile.c]: Could you please use the API? Both have one.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cfile.c (hunk 1 above)
Comment (XVilka) [reply librz/core/cfile.c]: See `rz_core_seek_opt(core, n, true, !silent);` and `RZ_API bool rz_core_debug_step_one(RzCore *core, int times)`

Comment (XVilka) [reply librz/core/cfile.c]: Or just `rz_core_seek(core, addr, false);`

File: librz/parse/p/parse_z80_pseudo.c [hunk 2]
```diff
@@ -11,76 +11,153 @@
...
+static RzList *z80_tokenize(const char *assembly, size_t length);
+
+// https://wikiti.brandonw.net/index.php?title=Z80_Instruction_Set
+static const RzPseudoGrammar z80_lexicon[] = {
+	RZ_PSEUDO_DEFINE_GRAMMAR("adc", "1 += 2 + cf"),
+	RZ_PSEUDO_DEFINE_GRAMMAR("add", "1 += 2"),
+	RZ_PSEUDO_DEFINE_GRAMMAR("and", "1 &= 2"),
+	RZ_PSEUDO_DEFINE_GRAMMAR("bit", "tmp = 2 & (#1 << 1)"),
+	RZ_PSEUDO_DEFINE_GRAMMAR("call", "call 1"),
+	//	RZ_PSEUDO_DEFINE_GRAMMAR("ccf", "1 = 1 ? 2"),
```
Comment (XVilka) [inline librz/parse/p/parse_z80_pseudo.c:25]: Why those are commented?

File: librz/parse/parse.c [hunk 3]
```diff
@@ -99,12 +99,24 @@ RZ_API bool rz_parse_assemble(RzParse *p, char *data, char *str) {
 }
 
 // data is input disasm, str is output pseudo
-// TODO: refactooring, this should return char * instead
-RZ_API bool rz_parse_parse(RzParse *p, const char *data, char *str) {
-	rz_return_val_if_fail(p && data && str, false);
-	return (p && data && *data && p->cur && p->cur->parse)
-		? p->cur->parse(p, data, str)
-		: false;
+RZ_API char *rz_parse_parse(RzParse *p, const char *data) {
```
Comment (XVilka) [inline librz/parse/parse.c]: I suggest to rename the function and give it more meaningful name. Also please add a Doxygen

**Review by XVilka — CHANGES_REQUESTED** [review]
Overall is good.

File: librz/parse/p/parse_z80_pseudo.c (hunk 2 above)
Comment (wargio) [reply librz/parse/p/parse_z80_pseudo.c:25]: those are the missing instructions

Comment (XVilka) [reply librz/parse/p/parse_z80_pseudo.c:25]: What do you mean by `missing`?

Comment (wargio) [reply librz/parse/p/parse_z80_pseudo.c:25]: not implemented in pseudo

Comment (XVilka) [reply librz/parse/p/parse_z80_pseudo.c:25]: ok.

**Review by XVilka — APPROVED** [review]

**Review by XVilka — APPROVED** [review]
Looks amazing. Lets see what others think and lets merge 👍 💯

**Comment by XVilka** [comment]
It also fixes one of the tests broken on PPC: https://github.com/rizinorg/rizin/pull/1250

**Comment by wargio** [comment]
yes it does.

**Pull request merged**